src/machines/java/com/enderio/machines/client/gui/widget/EnergyTextboxWidget.java

```diff
@@ -0,0 +1,124 @@
+package com.enderio.machines.client.gui.widget;
+
+import net.minecraft.client.gui.Font;
+import net.minecraft.client.gui.GuiGraphics;
+import net.minecraft.client.gui.components.EditBox;
+import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.network.chat.Component;
+import java.text.DecimalFormat;
+import java.text.NumberFormat;
+import java.util.Locale;
+import java.util.function.Supplier;
+
+public class EnergyTextboxWidget extends EditBox {
+    private final Supplier<Integer> maxEnergy;
+
+    protected final Screen displayOn;
+    
+    public EnergyTextboxWidget(Screen displayOn, Supplier<Integer> maxEnergy, Font font, int x, int y, int width, int height, Component message) {
+        super(font, x, y, width, height, message);
+        this.maxEnergy = maxEnergy;
+        this.displayOn = displayOn;
+        setCanLoseFocus(true);
+    }
+
+    //Input only accepts digits (0, 1, ..., 9)
+    @Override
+    public boolean charTyped(char codePoint, int modifiers) {
+
+        if (codePoint >= 48 && codePoint <= 57) {
+            int cursorPos = getCursorPosition() + 1;
+            boolean res = super.charTyped(codePoint, modifiers);
+            setValue(formatEnergy(getValue()));
+            moveCursorTo(cursorPos);
+
+            return res;
+        }
+
+        return false;
+    }
+
+    @Override
+    public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
+        boolean res = super.keyPressed(keyCode, scanCode, modifiers);
+
+        if (keyCode == 259) {
+            int cursorPos = getCursorPosition();
+            setValue(formatEnergy(getValue()));
+
+            if (getValue().equals("0")) {
+                moveCursorTo(1);
+            } else {
+                moveCursorTo(cursorPos);
+            }
+        }
+
+        return res;
+    }
+
+    @Override
+    public boolean mouseClicked(double mouseX, double mouseY, int button) {
+
+        if (mouseX >= getX() && mouseX <= getX() + width && mouseY >= getY() && mouseY <= getY() + height) {
+            super.mouseClicked(mouseX, mouseY, button);
+
+            //Clear content on right click
+            if (button == 1) {
+                setValue("0");
+                moveCursorTo(1);
+            }
+
+            return true;
+        } else {
+            setFocused(false);
+            return false;
+        }
+    }
+
+    public String formatEnergy(String value) {
+
+        if (value.isEmpty()) {
+            return "0";
+        }
+
+        int energy = 0;
+        value = value.replace(",", "");
+
+        try {
+            energy = Integer.parseInt(value);
+            energy = Math.min(energy, maxEnergy.get());
+        } catch(Exception e) {
+            energy = 0;
+        }
+
+        try {
+            DecimalFormat decimalFormat = (DecimalFormat) NumberFormat.getInstance(Locale.ROOT);
+            decimalFormat.applyPattern("#,###");
```

Comment:
What happens when we use big numbers like 1M? Maybe future proof it since the widget can be used somewhere else

Reply:
I tested this function in a normal java program just to see how the decimal format behaves and it works fine with bigger numbers